github.com/cockroachdb/pebble/internal/metamorphic: TestMetaTwoInstance failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13776604?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13776604?buildTab=artifacts#meta) on refs/heads/master @ [e53e09a9edbe](https://github.com/cockroachdb/pebble/commits/e53e09a9edbe):


```
        // INFO: [JOB 37] sstable deleted 000047
        // INFO: [JOB 37] sstable deleted 000049
        // INFO: [JOB 7] flushing: sstable created 000128
        // INFO: [JOB 37] sstable deleted 000050
        // INFO: [JOB 37] sstable deleted 000052
        // INFO: [JOB 37] sstable deleted 000053
        // INFO: [JOB 37] sstable deleted 000065
        // INFO: [JOB 37] sstable deleted 000066
        // INFO: [JOB 37] sstable deleted 000068
        // INFO: [JOB 37] sstable deleted 000069
        // INFO: [JOB 7] flushing: sstable created 000129
        // INFO: [JOB 37] sstable deleted 000070
        // INFO: [JOB 37] sstable deleted 000071
        // INFO: [JOB 37] sstable deleted 000072
        // INFO: [JOB 37] sstable deleted 000073
        // INFO: [JOB 37] sstable deleted 000082
        db1.Get("stjzvq@4") // [""] pebble: not found #1770
        // INFO: [JOB 37] sstable deleted 000083
        db1.Set("yqoycwbzp@1", "pahhbkgmlm") // <nil> #1771
        db1.Set("yrxcstgj@1", "knyagqjqmt") // <nil> #1774
        // INFO: [JOB 37] sstable deleted 000084
        db1.Delete("fatlkq@4") // <nil> #1778
        db1.Delete("cxcrwtekd@3") // <nil> #1781
        // INFO: [JOB 37] sstable deleted 000085
        db1.DeleteRange("bbtyvmqnut@6", "imbmqppeb@4") // <nil> #1782
        db1.Set("dolgaek@5", "k") // <nil> #1784
        // INFO: [JOB 7] flushing: sstable created 000130
        db1.DeleteRange("toquyqwflm@1", "xhnperzlxj@4") // <nil> #1786
        db1.Set("zaslhv", "qc") // <nil> #1789
        db1.RangeKeyDelete("fatlkq", "vffxhwlxvr") // <nil> #1791
        db1.Delete("voev@2") // <nil> #1796
        db1.Set("qguahklek", "dtjznojpu") // <nil> #1800
        db1.DeleteRange("fokurxk@5", "jytnccv@4") // <nil> #1807
        db1.Delete("istqjbw@2") // <nil> #1810
        db1.Set("caioijorn@1", "gteopry") // <nil> #1811
        db1.RangeKeySet("hjntenemx", "naahtmn", "@7", "lyojvhrx") // <nil> #1813
        // INFO: [JOB 7] flushing: sstable created 000131
        // INFO: [JOB 7] flushing: sstable created 000132
        // INFO: [JOB 7] flushing: sstable created 000133
        // INFO: [JOB 7] flushing: sstable created 000134
        // INFO: [JOB 7] flushing: sstable created 000135
        // INFO: [JOB 7] flushed 1 memtable (1.1KB) to L0 [000124 000125 000126 000127 000128 000129 000130 000131 000132 000133 000134 000135] (11KB), in 0.0s (0.0s total), output rate 1.4MB/s
        // INFO: [JOB 9] ingesting: sstable created 000136
        // INFO: [JOB 9] ingesting: sstable created 000137
        // INFO: [JOB 7] WAL deleted 000118
        // INFO: [JOB 10] flushing 1 ingested table
        // INFO: [JOB 10] flushed 1 ingested flushable L0:000121 (1.1KB) in 0.0s (0.0s total), output rate 25MB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240201-051930.646115143200/random-024 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-024 (0.81s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1706764770646603780 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13797796?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13797796?buildTab=artifacts#meta) on refs/heads/master @ [e53e09a9edbe](https://github.com/cockroachdb/pebble/commits/e53e09a9edbe):


```
        db2.Set("mbikrdcqj@9", "synrlx") // <nil> #4266
        db1.Delete("ljgdpizns@10") // <nil> #4281
        db1.DeleteRange("osnz@8", "waymyzem@3") // <nil> #4284
        snap31 = db2.NewSnapshot("jsphbfnrrk", "ncamijvns", "nwfnrztki", "pksdtlqwvoj", "rfrdihliexu", "zjizydv") #4286
        db1.Get("ojmnnih@9") // [""] pebble: not found #4287
        db1.DeleteRange("jqavdyf@1", "vktsdbo") // <nil> #4292
        db2.Set("fmofm@9", "ceaewspinfvzgjqtzul") // <nil> #4293
        db1.Delete("vmawyma@8") // <nil> #4304
        snap30.Get("gffy") // [""] pebble: not found #4303
        db2.Set("vljpapdzz@1", "hrjtbaedpzdr") // <nil> #4305
        snap31.Get("vmbguobhdm@5") // [""] pebble: not found #4307
        db2.Delete("mbikrdcqj@9") // <nil> #4308
        db2.DeleteRange("aaklijnpsuq@5", "rebknhakp@11") // <nil> #4320
        db1.Delete("gcejaglqm@2") // <nil> #4323
        db2.DeleteRange("jrelkxug@8", "mpksx@12") // <nil> #4325
        db1.Get("cgwxwe@6") // [""] pebble: not found #4340
        db2.Get("xbhju@3") // [""] pebble: not found #4342
        db1.Set("fmpgkytcyg@11", "ucmlftqwukkgpi") // <nil> #4346
        snap31.Get("rwvodduwih@4") // [""] pebble: not found #4350
        batch33 = db1.NewIndexedBatch() #4353
        batch33.Delete("schhnavrdj@6") // <nil> #4354
        batch33.Get("rgvbadowgl@3") // [""] pebble: not found #4355
        db2.Delete("zvlxjzmrq@6") // <nil> #4356
        batch33.Delete("xgjben@5") // <nil> #4358
        db2.Delete("yhjh@9") // <nil> #4360
        db2.Delete("dycyh@7") // <nil> #4361
        db1.Get("kyozywibckz@9") // [""] pebble: not found #4373
        // INFO: [JOB 25] ingesting: sstable created 000140
        // INFO: [JOB 26] WAL created 000141
        // INFO: [JOB 27] WAL created 000142
        // INFO: [JOB 25] ingested as flushable 000140 (1003B)
        db1.Ingest(batch33) // <nil> #4378
        db2.Set("khpnkfbxxej@11", "umhptjtgzbuft") // <nil> #4381
        db1.Delete("yifzvtunw@5") // <nil> #4382
        db2.Delete("qxhc@11") // <nil> #4383
        // INFO: [JOB 28] flushing 1 memtable (415B) to L0
        // INFO: [JOB 28] flushing: sstable created 000143
        snap30.Get("yifzvtunw@5") // [""] pebble: not found #4384
        snap32 = db1.NewSnapshot("gcjy", "ilpqqhihytf", "mrlp", "vljpapdzz") #4388
        db2.Delete("vmawyma@6") // <nil> #4392
        db2.Set("rfrdihliexu@8", "i") // <nil> #4393
        // INFO: [JOB 29] ingesting: sstable created 000144
        // INFO: [JOB 29] ingesting: sstable created 000145
        // INFO: [JOB 28] MANIFEST created 000146
        // INFO: [JOB 28] flushed 1 memtable (415B) to L0 [000143] (1.2KB), in 0.0s (0.0s total), output rate 165KB/s
        // INFO: [JOB 30] flushing 1 ingested table
        // INFO: [JOB 30] flushed 1 ingested flushable L0:000140 (1003B) in 0.0s (0.0s total), output rate 10MB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240202-050848.9391706257594/random-009 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-009 (0.88s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1706850528939952881 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13818195?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13818195?buildTab=artifacts#meta) on refs/heads/master @ [087fd92aca44](https://github.com/cockroachdb/pebble/commits/087fd92aca44):


```
        db2.Delete("rkmzf@4") // <nil> #291
        db2.Get("vyntee@1") // [""] pebble: not found #301
        db2.Delete("pzusa") // <nil> #312
        db2.DeleteRange("lgkr@3", "zryegv@3") // <nil> #318
        db2.DeleteRange("lgkr@3", "mhplcqvujml@1") // <nil> #321
        db2.Get("kmgvp@2") // [""] pebble: not found #331
        db2.RangeKeySet("ucgufxk", "zryegv", "@6", "fxfotqzkwqdzbnca") // <nil> #339
        db2.Get("mzqiovd") // [""] pebble: not found #340
        batch5 = db2.NewBatch() #344
        batch5.Set("zryegv@6", "nkyvlfletqbp") // <nil> #347
        iter5.First() // [true,"lgkr",<no point>,["lgkr","lgkr@1")=>{"@4"="kq","@2"="btxjjxmnrvmua"}*] <nil> #349
        iter3.Prev("") // [false] <nil> #350
        iter1.SetBounds("zryegv@6", "zryegv@3") // <nil> #351
        iter1.SeekLT("zryegv@3", "") // [false] <nil> #352
        iter1.Next("") // [false] <nil> #353
        iter1.Prev("") // [false] <nil> #354
        iter3.SetBounds("vyntee@1", "zryegv@6") // <nil> #355
        iter3.Next("") // [false] <nil> #356
        db1.Get("zryegv@3") // ["goriv"] <nil> #357
        iter3.SeekPrefixGE("axkwucl@4") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #360
        iter1.SeekLT("pzusa", "") // [false] <nil> #362
        db1.Get("vyntee@3") // [""] pebble: not found #363
        iter3.SeekLT("rkmzf", "") // [false] <nil> #364
        iter3.SeekLT("rkmzf@2", "") // [false] <nil> #365
        iter3.SeekLT("ucgufxk@1", "kmgvp@2") // [invalid] <nil> #366
        iter1.SeekLT("vyntee@1", "") // [false] <nil> #367
        db1.RangeKeyUnset("axkwucl", "jnhvymugjvh", "@2") // <nil> #368
        iter1.Last() // [false] <nil> #369
        db1.Delete("ucgufxk@1") // <nil> #370
        iter3.SetBounds("lgkr@3", "vyntee@2") // <nil> #372
        db1.Set("ovrlojl@3", "bbenptcvxey") // <nil> #373
        iter1.SeekGE("zryegv@6", "zryegv@3") // [invalid] <nil> #374
        iter3.Next("") // [true,"lgkr@3",<no point>,["lgkr@3","mhplcqvujml")=>{"@4"="kq","@2"="btxjjxmnrvmua"}*] <nil> #376
        // INFO: [JOB 17] ingesting: sstable created 000018
        // INFO: [JOB 18] WAL created 000019
        // INFO: [JOB 19] WAL created 000020
        // INFO: [JOB 17] ingested as flushable 000018 (914B)
        db2.Ingest(batch5) // <nil> #348
        snap2 = db2.NewSnapshot("rkmzf", "vyntee") #358
        db2.Set("jnhvymugjvh@3", "o") // <nil> #359
        snap2.Get("rkmzf@2") // [""] pebble: not found #361
        db2.Get("jkyg@2") // [""] pebble: not found #371
        db2.Delete("ziihtaqrsq@4") // <nil> #375
        // INFO: [JOB 20] flushing 1 memtable (529B) to L0
        // INFO: [JOB 20] flushing: sstable created 000021
        // INFO: [JOB 21] ingesting: sstable created 000022
        // INFO: [JOB 20] flushed 1 memtable (529B) to L0 [000021] (1.2KB), in 0.0s (0.0s total), output rate 4.2MB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240203-053207.7071501123197/random-014 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-014 (0.09s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1706938327707129124 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/13818196?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/13818196?buildTab=artifacts#meta) on refs/heads/master @ [087fd92aca44](https://github.com/cockroachdb/pebble/commits/087fd92aca44):


```
        iter18.Next("") // [false] <nil> #2008
        // INFO: [JOB 38] ingested as flushable 000037 (1023B)
        db1.Ingest(batch14) // <nil> #1753
        // INFO: [JOB 36] flushed 1 ingested flushable L0:000029 (1.1KB) in 0.0s (0.1s total), output rate 468KB/s
        db1.Delete("gotc@1") // <nil> #1756
        // INFO: [JOB 39] flushing 1 memtable (726B) to L0
        iter19.Last() // [false] <nil> #1763
        iter19.SetBounds("gyhhosu@6", "iwzyumrbet@9") // <nil> #1766
        iter19.Next("wqizhk") // [invalid] <nil> #1767
        db1.Delete("xmspa@1") // <nil> #1779
        db1.Get("xmspa@1") // [""] pebble: not found #1780
        db1.Set("wqizhk", "") // <nil> #1791
        iter19.First() // [false] <nil> #1793
        iter19.SeekLT("vageuujcnz@3", "") // [false] <nil> #1794
        // INFO: [JOB 39] flushing: sstable created 000038
        // INFO: [JOB 39] flushing: sstable created 000039
        // INFO: [JOB 39] flushing: sstable created 000040
        // INFO: [JOB 39] flushing: sstable created 000041
        // INFO: [JOB 39] flushing: sstable created 000042
        snap16 = db1.NewSnapshot("cpqplevqj", "gmjdx", "igubjrnol", "qahbuofa") #1795
        iter19.SetOptions("", "iwzyumrbet@9", 2 /* key types */, 0, 0, false /* use L6 filters */, "@4" /* masking suffix */) // <nil> #1796
        iter19.SeekPrefixGE("vageuujcnz@3") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #1797
        iter19.Last() // [true,"fonxr",<no point>,["fonxr","ghojcitxo")=>{"@5"="qjhwwnklr"}*] <nil> #1808
        db1.Get("kfzaislfcma@1") // [""] pebble: not found #1809
        iter19.Last() // [true,"fonxr",<no point>,["fonxr","ghojcitxo")=>{"@5"="qjhwwnklr"}] <nil> #1810
        iter19.First() // [true,"byvjdybs@7","caaksbokokpavuuk",<no range>*] <nil> #1813
        iter19.SetBounds("nobgehzljek@2", "offwn@6") // <nil> #1814
        iter19.SetBounds("cpqplevqj@9", "fonxr@1") // <nil> #1816
        iter19.Last() // [true,"fonxr",<no point>,["fonxr","fonxr@1")=>{"@5"="qjhwwnklr"}*] <nil> #1820
        iter19.First() // [true,"dfuy@3","gkoggudzz",<no range>*] <nil> #1821
        // INFO: [JOB 37] compacted(default) L0 [000023 000022 000030 000027 000028 000026 000034 000035 000033 000031 000032] (14KB) Score=111.11 + L6 [000021] (1.4KB) Score=0.01 -> L6 [000036] (1.6KB), in 0.1s (0.1s total), output rate 27KB/s
        // INFO: [JOB 39] flushing: sstable created 000043
        // INFO: [JOB 37] sstable deleted 000026
        // INFO: [JOB 37] sstable deleted 000027
        // INFO: [JOB 37] sstable deleted 000028
        // INFO: [JOB 37] sstable deleted 000030
        // INFO: [JOB 37] sstable deleted 000031
        // INFO: [JOB 37] sstable deleted 000032
        // INFO: [JOB 37] sstable deleted 000033
        // INFO: [JOB 37] sstable deleted 000034
        // INFO: [JOB 37] sstable deleted 000035
        // INFO: [JOB 39] flushed 1 memtable (726B) to L0 [000038 000039 000040 000041 000042 000043] (7.1KB), in 0.0s (0.0s total), output rate 383KB/s
        // INFO: [JOB 40] flushing 1 ingested table
        // INFO: [JOB 40] flushed 1 ingested flushable L0:000037 (1023B) in 0.0s (0.0s total), output rate 4.0MB/s
        // INFO: [JOB 42] ingesting: sstable created 000044
        // INFO: [JOB 42] ingesting: sstable created 000045
        // INFO: [JOB 42] ingesting: sstable created 000046
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240203-053751.6461988617153/random-000 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-000 (2.15s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1706938671646193504 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13830469?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13830469?buildTab=artifacts#meta) on refs/heads/master @ [087fd92aca44](https://github.com/cockroachdb/pebble/commits/087fd92aca44):


```
        iter59.First() // [true,"cjng@14","dwboguffglwmlmp",<no range>] <nil> #5896
        iter59.InternalNext() // <nil> #5897
        iter59.SetOptions("", "bsjzcjdmkj@6", 0 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */) // <nil> #5898
        iter59.First() // [true,"aymjcmmxv@5","ihqjlecxjvlgaaxbyg",<no range>] <nil> #5899
        iter59.First() // [true,"aymjcmmxv@5","ihqjlecxjvlgaaxbyg",<no range>] <nil> #5902
        iter59.Next("yhyethovswr@4") // [invalid] <nil> #5903
        iter59.Prev("") // [true,"aymjcmmxv@5","ihqjlecxjvlgaaxbyg",<no range>] <nil> #5904
        iter59.SeekPrefixGE("neusjnlqxwv@7") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #5905
        iter59.Prev("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #5906
        iter59.Prev("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #5908
        // INFO: [JOB 6] flushed 1 memtable (720B) to L0 [000256] (1.2KB), in 0.0s (0.0s total), output rate 1.5MB/s
        // INFO: [JOB 7] flushing 1 ingested table
        // INFO: [JOB 9] flushed 1 memtable (107B) to L0 [000206] (976B), in 0.0s (0.0s total), output rate 1.1MB/s
        // INFO: [JOB 6] WAL deleted 000250
        // INFO: [JOB 9] WAL deleted 000202
        // INFO: [JOB 9] MANIFEST deleted 000196
        // INFO: [JOB 8] ingesting: sstable created 000257
        // INFO: [JOB 7] MANIFEST created 000216
        // INFO: [JOB 7] flushed 1 ingested flushable L0:000253 (968B) in 0.0s (0.0s total), output rate 25MB/s
        // INFO: [JOB 7] WAL deleted 000254
        // INFO: [JOB 7] sstable deleted 000206
        // INFO: [JOB 7] MANIFEST deleted 000199
        // INFO: [JOB 7] ingested L6:000204 (935B), L0:000208 (107B), L0:000209 (69B), L0:000210 (119B), L0:000211 (120B), L0:000212 (36B), L0:000213 (121B), L6:000214 (206B), L6:000215 (199B)
        db2.IngestAndExcise(batch43, "jivjn", "uocqc") // <nil> #5827
        db2.Set("tqmcrqcmv@3", "czxvyyuldfg") // <nil> #5828
        // INFO: [JOB 12] validated table: 000204:[uksbfvvsqb@6#663,0-uksbfvvsqb@6#663,0]
        db2.Set("yted@12", "ytsaumhzzdrqrmjlfoj") // <nil> #5834
        db2.DeleteRange("fwmdtan@4", "tqmcrqcmv@2") // <nil> #5838
        db2.Set("uphogahk@12", "cortc") // <nil> #5851
        // INFO: [JOB 12] validated table: 000208:[badptkx@1#628,0-jivjn#72057594037927935,15]
        db2.Get("okfjtlkvq@6") // [""] pebble: not found #5855
        db2.RangeKeySet("dkxfjv", "hhfjtxkjru", "@12", "dggwoabycldriqs") // <nil> #5858
        // INFO: [JOB 12] validated table: 000210:[badptkx@1#635,15-jivjn#72057594037927935,15]
        db2.Get("zvuqhd@14") // ["fjju"] <nil> #5869
        db2.Delete("tuirg@6") // <nil> #5889
        db2.Set("iytcbwkfl@15", "hzom") // <nil> #5893
        db2.DeleteRange("leqmukcjbc@9", "osxzdwg@13") // <nil> #5900
        iter59.Close() // pebble: SeekPrefixGE supplied with key outside of upper bound #5909
        db2.RangeKeyUnset("okfjtlkvq", "orronuqxk", "@6") // <nil> #5910
        // INFO: [JOB 13] WAL created 000217
        // INFO: [JOB 12] validated table: 000212:[baxvhjxbgp@1#657,15-jivjn#72057594037927935,15]
        // INFO: [JOB 14] flushing 1 memtable (543B) to L0
        // INFO: [JOB 12] validated table: 000214:[aeydzhxfjf#536,21-jivjn#72057594037927935,21]
        // INFO: [JOB 14] flushing: sstable created 000218
        // INFO: [JOB 15] compacting(elision-only) L6 [000204] (935B) Score=0.00 + L6 [] (0B) Score=0.00; OverlappingRatio: Single 0.00, Multi 0.00
        // INFO: [JOB 15] MANIFEST created 000219
        // INFO: [JOB 15] compacted(elision-only) L6 [000204] (935B) Score=0.00 + L6 [] (0B) Score=0.00 -> L6 [] (0B), in 0.0s (0.0s total), output rate 0B/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240204-054458.7642089707422/random-003 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-003 (0.99s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1707025498764819037 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13840805?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicTwoInstance/13840805?buildTab=artifacts#meta) on refs/heads/master @ [087fd92aca44](https://github.com/cockroachdb/pebble/commits/087fd92aca44):


```
        db1.RangeKeyUnset("scuyk", "yjmt", "@2") // <nil> #103
        db1.Set("hseig@1", "bgwiylsfeal") // <nil> #106
        db1.RangeKeyUnset("fuzd", "kcdjhoetlcf", "") // <nil> #107
        db1.Delete("yjmt") // <nil> #108
        snap1.Get("ugxtnxm@2") // [""] pebble: not found #110
        // INFO: [JOB 11] WAL created 000013
        db2.Delete("ftbl@1") // <nil> #104
        snap3 = db2.NewSnapshot("scuyk", "ugxtnxm", "uqtiamapxtx", "xjtxxooor") #105
        iter1 = snap3.NewIter("ufdydef", "ufhltbn", 2 /* key types */, 1, 2, false /* use L6 filters */, "" /* masking suffix */) // <nil> #109
        iter1.Next("") // [false] <nil> #111
        iter1.SetBounds("tovnhil", "ttpzowl") // <nil> #112
        iter1.SeekLT("ttpzowl", "") // [false] <nil> #113
        iter1.Prev("") // [false] <nil> #114
        db1.Delete("miwfmmdir@2") // <nil> #115
        iter1.SeekLT("uqtiamapxtx@1", "") // [false] <nil> #116
        iter1.First() // [false] <nil> #117
        db1.Set("xzgujtonm@2", "bdcwfco") // <nil> #118
        iter1.Next("") // [false] <nil> #119
        iter1.SeekGE("uqtiamapxtx@1", "") // [false] <nil> #120
        iter1.SeekPrefixGE("wcxxfqnwsvg") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #121
        iter1.Prev("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #123
        iter1.SeekPrefixGE("xzgujtonm@2") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #124
        iter1.SeekPrefixGE("xjtxxooor@2") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #125
        iter1.SeekLT("fuzd@2", "") // [false] <nil> #126
        iter1.Last() // [false] <nil> #127
        iter1.SeekPrefixGE("ugxtnxm@2") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #128
        iter1.SeekLT("ugxtnxm", "") // [false] <nil> #129
        iter1.SetBounds("wszbxjcpqwy", "wtezswhghrx") // <nil> #130
        iter1.SeekGE("wszbxjcpqwy", "") // [false] <nil> #131
        iter1.Next("") // [false] <nil> #132
        iter1.Next("") // [false] <nil> #133
        iter1.SeekLT("hseig@1", "") // [false] <nil> #135
        iter1.First() // [false] <nil> #136
        batch2 = db1.NewIndexedBatch() #137
        iter1.SetBounds("vljmtowtwyu", "vtcokzsydyn") // <nil> #139
        iter1.Next("") // [false] <nil> #141
        // INFO: [JOB 12] flushing 1 memtable (293B) to L0
        // INFO: [JOB 12] flushing: sstable created 000014
        // INFO: [JOB 12] flushed 1 memtable (293B) to L0 [000014] (1.2KB), in 0.0s (0.0s total), output rate 1.8MB/s
        // INFO: [JOB 12] WAL deleted 000010
        db2.Delete("fuzd@2") // <nil> #122
        db2.Set("kcdjhoetlcf@1", "wqjsqnaxr") // <nil> #134
        db2.Set("miwfmmdir", "ixyuyzehzczdvklj") // <nil> #138
        batch2.Set("hseig@1", "yo") // <nil> #140
        batch2.Delete("pvjuhqpvyj@1") // <nil> #142
        // INFO: [JOB 14] ingesting: sstable created 000015
        // INFO: [JOB 14] ingesting: sstable created 000016
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240205-050614.471479073461/random-020 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-020 (0.39s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1707109574471523250 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/13840806?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/13840806?buildTab=artifacts#meta) on refs/heads/master @ [087fd92aca44](https://github.com/cockroachdb/pebble/commits/087fd92aca44):


```
        // INFO: [JOB 1563] flushing: sstable created 001199
        // INFO: [JOB 1555] sstable deleted 001140
        // INFO: [JOB 1555] sstable deleted 001171
        // INFO: [JOB 1563] flushing: sstable created 001200
        // INFO: [JOB 1557] sstable deleted 001163
        // INFO: [JOB 1557] sstable deleted 001176
        // INFO: [JOB 1563] flushing: sstable created 001201
        // INFO: [JOB 1558] sstable deleted 001131
        // INFO: [JOB 1558] sstable deleted 001150
        // INFO: [JOB 1563] flushing: sstable created 001202
        // INFO: [JOB 1558] sstable deleted 001170
        // INFO: [JOB 1558] sstable deleted 001177
        // INFO: [JOB 1563] flushing: sstable created 001203
        // INFO: [JOB 1560] sstable deleted 001180
        // INFO: [JOB 1560] sstable deleted 001181
        // INFO: [JOB 1562] sstable deleted 001179
        // INFO: [JOB 1563] flushing: sstable created 001204
        // INFO: [JOB 1562] sstable deleted 001182
        // INFO: [JOB 1563] flushing: sstable created 001205
        // INFO: [JOB 1563] flushing: sstable created 001206
        // INFO: [JOB 1563] flushing: sstable created 001207
        // INFO: [JOB 1322] compacting(move) L2 [001111] (946B) Score=15.79 + L3 [] (0B) Score=8.51; OverlappingRatio: Single 0.00, Multi 0.00
        // INFO: [JOB 1306] sstable deleted 000972
        // INFO: [JOB 1306] sstable deleted 000976
        // INFO: [JOB 1306] sstable deleted 000979
        // INFO: [JOB 1306] sstable deleted 000980
        // INFO: [JOB 1306] sstable deleted 000981
        // INFO: [JOB 1306] sstable deleted 000983
        // INFO: [JOB 1306] sstable deleted 000986
        // INFO: [JOB 1306] sstable deleted 000990
        // INFO: [JOB 1306] sstable deleted 000992
        // INFO: [JOB 1306] sstable deleted 000993
        // INFO: [JOB 1306] sstable deleted 000994
        // INFO: [JOB 1306] sstable deleted 000995
        // INFO: [JOB 1306] sstable deleted 000996
        // INFO: [JOB 1306] sstable deleted 000997
        // INFO: [JOB 1306] sstable deleted 000998
        // INFO: [JOB 1306] sstable deleted 000999
        // INFO: [JOB 1306] sstable deleted 001000
        // INFO: [JOB 1306] sstable deleted 001002
        // INFO: [JOB 1306] sstable deleted 001003
        // INFO: [JOB 1306] sstable deleted 001004
        // INFO: [JOB 1320] compacted(move) multilevelL4 [001104] (1017B) Score=12.83 + L5 [] (0B) Score=4.26 + L6 [] (0B) Score=0.43 -> L6 [001104] (1017B), in 0.0s (0.0s total), output rate 53MB/s
        // INFO: [JOB 1323] compacting(move) L3 [001107] (1.3KB) Score=12.64 + L4 [] (0B) Score=9.09; OverlappingRatio: Single 0.00, Multi 0.00
        // INFO: [JOB 1306] sstable deleted 001005
        // INFO: [JOB 1322] compacted(move) L2 [001111] (946B) Score=15.79 + L3 [] (0B) Score=8.51 -> L3 [001111] (946B), in 0.0s (0.0s total), output rate 29MB/s
        // INFO: [JOB 1563] flushed 1 memtable (2.9KB) to L0 [001185 001186 001188 001189 001190 001191 001193 001194 001195 001196 001197 001198 001199 001200 001201 001202 001203 001204 001205 001206 001207] (23KB), in 0.0s (0.1s total), output rate 509KB/s
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/240205-052129.4153616647710/random-016 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/random-016 (28.83s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1707110489415595128 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/compaction.go
+++ b/compaction.go
@@ -2060,7 +2060,35 @@ func (d *DB) flush1() (bytesFlushed uint64, err error) {
 				s = s.next
 				continue
 			}
+			overlapsFlushable := false
 			if base.Visible(earliestUnflushedSeqNum, s.efos.seqNum, InternalKeySeqNumMax) {
+				// There are some unflushed keys that are still visible to the EFOS.
+				// Check if any memtables older than the EFOS contain keys within a
+				// protected range of the EFOS. If no, we can transition.
+				protectedRanges := make([]bounded, len(s.efos.protectedRanges))
+				for i := range s.efos.protectedRanges {
+					protectedRanges[i] = s.efos.protectedRanges[i]
+				}
+				for i := range d.mu.mem.queue {
+					if !base.Visible(d.mu.mem.queue[i].logSeqNum, s.efos.seqNum, InternalKeySeqNumMax) {
+						// All keys in this memtable are newer than the EFOS. Skip this
+						// memtable.
+						continue
+					}
+					// NB: computePossibleOverlaps could have false positives, such as if
+					// the flushable is a flushable ingest and not a memtable. In that
+					// case we don't open the sstables to check; we just pessimistically
+					// assume an overlap.
+					d.mu.mem.queue[i].computePossibleOverlaps(func(b bounded) shouldContinue {
+						overlapsFlushable = true
+						return stopIteration
+					}, protectedRanges...)
+					if overlapsFlushable {
+						break
+					}
+				}
+			}
+			if overlapsFlushable {
 				s = s.next
 				continue
 			}

--- a/ingest.go
+++ b/ingest.go
@@ -1570,7 +1570,7 @@ func (d *DB) ingest(
 
 		// Assign the sstables to the correct level in the LSM and apply the
 		// version edit.
-		ve, err = d.ingestApply(jobID, loadResult, targetLevelFunc, mut, exciseSpan)
+		ve, err = d.ingestApply(jobID, loadResult, targetLevelFunc, mut, exciseSpan, seqNum)
 	}
 
 	// Only one ingest can occur at a time because if not, one would block waiting
@@ -2082,6 +2082,7 @@ func (d *DB) ingestApply(
 	findTargetLevel ingestTargetLevelFunc,
 	mut *memTable,
 	exciseSpan KeyRange,
+	exciseSeqNum uint64,
 ) (*versionEdit, error) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
@@ -2310,33 +2311,36 @@ func (d *DB) ingestApply(
 				}
 			}
 		}
-		// Check for any EventuallyFileOnlySnapshots that could be watching for
-		// an excise on this span. There should be none as the
-		// computePossibleOverlaps steps should have forced these EFOS to transition
-		// to file-only snapshots by now. If we see any that conflict with this
-		// excise, panic.
-		if exciseSpan.Valid() {
-			for s := d.mu.snapshots.root.next; s != &d.mu.snapshots.root; s = s.next {
-				if s.efos == nil {
-					continue
-				}
-				efos := s.efos
-				// TODO(bilal): We can make this faster by taking advantage of the sorted
-				// nature of protectedRanges to do a sort.Search, or even maintaining a
-				// global list of all protected ranges instead of having to peer into every
-				// snapshot.
-				for i := range efos.protectedRanges {
-					if efos.protectedRanges[i].OverlapsKeyRange(d.cmp, exciseSpan) {
-						panic("unexpected excise of an EventuallyFileOnlySnapshot's bounds")
-					}
-				}
-			}
-		}
 	}
 	if err := d.mu.versions.logAndApply(jobID, ve, metrics, false /* forceRotation */, func() []compactionInfo {
 		return d.getInProgressCompactionInfoLocked(nil)
 	}); err != nil {
 		return nil, err
+	}
+
+	// Check for any EventuallyFileOnlySnapshots that could be watching for
+	// an excise on this span. There should be none as the
+	// computePossibleOverlaps steps should have forced these EFOS to transition
+	// to file-only snapshots by now. If we see any that conflict with this
+	// excise, panic.
+	if exciseSpan.Valid() {
+		for s := d.mu.snapshots.root.next; s != &d.mu.snapshots.root; s = s.next {
+			// Skip non-EFOS snapshots, and also skip any EFOS that were created
+			// *after* the excise.
+			if s.efos == nil || base.Visible(exciseSeqNum, s.efos.seqNum, base.InternalKeySeqNumMax) {
+				continue
+			}
+			efos := s.efos
+			// TODO(bilal): We can make this faster by taking advantage of the sorted
+			// nature of protectedRanges to do a sort.Search, or even maintaining a
+			// global list of all protected ranges instead of having to peer into every
+			// snapshot.
+			for i := range efos.protectedRanges {
+				if efos.protectedRanges[i].OverlapsKeyRange(d.cmp, exciseSpan) {
+					panic("unexpected excise of an EventuallyFileOnlySnapshot's bounds")
+				}
+			}
+		}
 	}
 
 	d.mu.versions.metrics.Ingest.Count++

